### PR TITLE
Move states of search to coordinating node

### PR DESCRIPTION
**Pull request by dnhatn**
This commit moves the states of search to the coordinating node instead of keeping them in the data node. Changes in this commit:

1. Include ShardSearchRequest to QuerySearchResult
2. Resend ShardSearchRequest in step 1 in the fetch phase so we can recreate a SearchContext from ShardFetchRequest
3. Include AggregatedDfs to ShardFetchRequest as we need it in both the query and fetch phase
4. Include RecoreDocIds to QuerySearchResult and send it back in the fetch phase

Relates #46523

I ran a few benchmarks for this change on a single node with http_logs, geonames, and geopoint. There was no regression. I will try to benchmark it with multiple nodes.

### CODE REVIEW

**Comment by dnhatn** [comment]
> testHistoryIsWrittenWithFailure

@elasticmachine test this please

File: modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorQuerySearchIT.java [hunk 1]
```diff
@@ -71,6 +75,14 @@
...
 public class PercolatorQuerySearchIT extends ESIntegTestCase {
 
+    @Override
+    protected Collection<Class<? extends Plugin>> nodePlugins() {
+        List<Class<? extends Plugin>> plugins = new ArrayList<>();
+        plugins.addAll(super.nodePlugins());
+        plugins.add(PercolatorPlugin.class);
+        return plugins;
+    }
+
```
Comment (jimczi) [inline modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorQuerySearchIT.java]: Is this change needed for this pr ? It seems unrelated.

File: server/src/main/java/org/elasticsearch/search/SearchPhaseResult.java [hunk 2]
```diff
@@ -90,6 +94,23 @@ public QuerySearchResult queryResult() {
...
+        return rescoreDocIds;
+    }
+
+    public void setRescoreDocIds(RescoreDocIds rescoreDocIds) {
+        this.rescoreDocIds = rescoreDocIds;
+    }
+
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         // TODO: this seems wrong, SearchPhaseResult should have a writeTo?
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/search/SearchPhaseResult.java]: we could apply the TODO here and clean the (de)serialization ? This would mean reconciling the logic between the different implementations and move the serialization of the object accessible here to this function. Happy to do this in a follow up though.

**Review by jimczi — APPROVED** [review]
I really like the approach here. The backward compatibility layer is simple and efficient since the main logic is unchanged (we rebuild the search context in all cases). I also find it nice that we send the rewritten shard request back and forth between phases. This effectively moves the state of the search to the coordinating node. 
I left one comment that could be addressed in a follow up. The rest looks good to me.

File: modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorQuerySearchIT.java (hunk 1 above)
Comment (dnhatn) [reply modules/percolator/src/test/java/org/elasticsearch/percolator/PercolatorQuerySearchIT.java]: The problem is that MockNioTransport does not register [NamedWriteables](https://github.com/elastic/elasticsearch/blob/95d46332f8c5e52e731fabcfee1c78056f0fa38d/test/framework/src/main/java/org/elasticsearch/test/transport/MockTransportService.java#L116) from the [bundled plugins](https://github.com/elastic/elasticsearch/blob/95d46332f8c5e52e731fabcfee1c78056f0fa38d/server/src/main/java/org/elasticsearch/node/Node.java#L400). We need to manually register the percolate plugin in the test as we now de/serialize the Percolate QueryBuilder between search phases. I looked into how to fix this issue, but it should be in a follow-up.

File: server/src/main/java/org/elasticsearch/search/SearchPhaseResult.java (hunk 2 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/SearchPhaseResult.java]: ++. I will do this in a follow-up.

**Comment by dnhatn** [comment]
@jimczi Thank you for your review.
